packages/inter-protocol/src/proposals/startPSM.js

```diff
@@ -18,24 +20,115 @@ import {
+ * Decode vstorage value to CapData
+ * XXX already written somewhere?
+ *
+ * @param {unknown} value
+ */
+const decodeToCapData = value => {
+  assert.typeof(value, 'string');
+
+  // { blockHeight: 123, values: [ ... ] }
+  const item = JSON.parse(value); // or throw
+  assert.typeof(item, 'object');
+  assert(item);
+  const { values } = item;
+  assert(Array.isArray(values));
+
+  assert.equal(values.length, 1);
+  // { body: "...", slots: [ ... ] }
+  const data = JSON.parse(values[0]);
+  assert.typeof(data, 'object');
+  assert(data);
+  assert.typeof(data.body, 'string');
+  assert(Array.isArray(data.slots));
+
+  /** @type {import('@endo/marshal').CapData<string>} */
+  // @ts-expect-error cast
+  const capData = data;
+  return capData;
+};
+
+/**
+ * Provide access to object graphs serialized in vstorage.
+ *
+ * @param {Array<[string, string]>} entries
+ * @param {(slot: string, iface?: string) => any} [slotToVal]
+ */
+export const makeHistoryReviver = (entries, slotToVal = undefined) => {
+  const board = boardSlottingMarshaller(slotToVal);
+  const vsMap = makeMap(entries);
+
+  const getItem = key => {
+    const raw = vsMap.get(key) || Fail`no ${key}`;
+    const capData = decodeToCapData(raw);
+    return harden(board.fromCapData(capData));
+  };
+  const children = prefix => [
+    ...makeSet(
+      entries
+        .map(([k, _]) => k)
+        .filter(k => k.length > prefix.length && k.startsWith(prefix))
+        .map(k => k.slice(prefix.length).split('.')[0]),
+    ),
+  ];
+  return harden({ getItem, children, has: k => vsMap.has(k) });
+};
+
+/**
+ * @param {Array<[key: string, value: string]>} chainStorageEntries
+ * @param {string} keyword
+ * @param {{ minted: Brand<'nat'>, anchor: Brand<'nat'> }} brands
+ * @returns {{ metrics?: MetricsNotification, governance?: Record<string, *> }}
+ */
+const findOldPSMState = (chainStorageEntries, keyword, brands) => {
+  // In this reviver, object references are revived as boardIDs
+  // from the pre-bulldozer board.
+  const toSlotReviver = makeHistoryReviver(chainStorageEntries);
+  if (!toSlotReviver.has(`published.psm.${Stable.symbol}.${keyword}.metrics`)) {
+    return {};
+  }
+  const metricsWithOldBoardIDs = toSlotReviver.getItem(
+    `published.psm.${Stable.symbol}.${keyword}.metrics`,
+  );
+  const oldIDtoNewBrand = makeMap([
+    [metricsWithOldBoardIDs.feePoolBalance.brand, brands.minted],
+    [metricsWithOldBoardIDs.anchorPoolBalance.brand, brands.anchor],
+  ]);
+  // revive brands; other object references map to undefined
+  const brandReviver = makeHistoryReviver(chainStorageEntries, s =>
+    oldIDtoNewBrand.get(s),
+  );
+  return {
+    metrics: brandReviver.getItem(`published.psm.IST.${keyword}.metrics`),
+    governance: brandReviver.getItem(`published.psm.IST.${keyword}.governance`)
+      .current,
+  };
+};
+
+/**
+ * @typedef {{
+ *   vatParameters: { chainStorageEntries?: Array<[k: string, v: string]>,
+ * }}} ChainStorageVatParams
```

Comment:
this type is not a concern of inter-protocol. I think it belongs in swingset-vats or cosmic-swingset.

Reply:
`vats/core/types.js` has a collection of types that swingset should export: #4655

I moved this one with those.